msprime/mutations.py

```diff
@@ -63,6 +69,113 @@ def asdict(self):
             self.__init__).parameters.keys() if hasattr(self, key)}
 
 
+class FiniteSites(object):
+    """
+    """
+    def __init__(self, seed, rate_map, alphabet, mutation_matrix, alphabet_distribution):
+        if alphabet not in [BINARY, NUCLEOTIDES]:
+            raise ValueError("Bad alphabet")
+        seed = int(seed)
+        rng = _msprime.RandomGenerator(seed)
+        alphabet_length = {BINARY: 2, NUCLEOTIDES: 4}[alphabet]
+        try:
+            valid_input = (
+                    len(mutation_matrix) == alphabet_length
+                    and all(len(x) == alphabet_length for x in mutation_matrix)
+                    and all(sum(x) == 1.0 for x in mutation_matrix)
+                    and all(all(u >= 0 for u in x) for x in mutation_matrix)
+                    and len(alphabet_distribution) == alphabet_length
+                    and sum(alphabet_distribution) == 1.0
+                    and all(x >= 0 for x in alphabet_distribution)
+                    )
+            if not valid_input:
+                raise ValueError("Illegal mutation matrix: must be a square matrix"
+                                 " with rows that sum to 1.0.")
+        except TypeError:
+            raise ValueError("Illegal mutation matrix: must be a square matrix"
+                             " with rows that sum to 1.0.")
+        self.alphabet = alphabet
+        self.mutation_matrix = mutation_matrix
+        self.mutation_generator = FiniteSitesMutationGenerator(
+            rng=rng, rate_map=rate_map, alphabet=alphabet,
+            mutation_matrix=mutation_matrix, alphabet_distribution=alphabet_distribution)
+
+    def asdict(self):
+        """
+        Returns a dict of arguments to recreate this class
+        """
+        return {key: getattr(self, key) for key in inspect.signature(
+            self.__init__).parameters.keys() if hasattr(self, key)}
+
+
+class FiniteSitesMutationGenerator(object):
+    """
+    """
+    def __init__(self, rng, rate_map, alphabet, mutation_matrix, alphabet_distribution):
+        self.rng = rng
+        self.rate_map = rate_map
+        self.alphabet = alphabet
+        self.mutation_matrix = mutation_matrix
+        self.alphabet_distribution = alphabet_distribution
+
+    def pick_allele(self):
+        u = np.random.uniform()
```

Comment:
We should either use our own rng GSL instance for this, or create a local numpy.random instance which we seed (from the GSL rng I guess?). We definitely shouldn't be calling the global numpy rng and altering its state though.